pulsar-client/src/main/java/org/apache/pulsar/client/impl/PartitionedConsumerImpl.java

```diff
@@ -191,7 +200,9 @@ protected Message internalReceive(int timeout, TimeUnit unit) throws PulsarClien
         Message message;
         try {
             lock.writeLock().lock();
-            message = incomingMessages.poll(0, TimeUnit.SECONDS);
+            message = (incomingMessages.size() > 0) ?
+                    incomingMessages.poll(0, TimeUnit.SECONDS) : messageFromConsumerImpl();
```

Comment:
I think it will not be necessary because it does the similar thing at line#209 by calling `resumeReceivingFromPausedConsumersIfNeeded()`. So, if queue is empty then `resumeReceivingFromPausedConsumersIfNeeded()` try to receive message from next consumer which has message available in the queue.

Reply:
Great! You are right, but isn't `internalReceiveAsync` being called for `receiveAsync`? In such case, if `incomingMessages` is empty `resumeReceivingFromPausedConsumersIfNeeded` won't be called and there could be some messages in consumers. 

I think just changing the if condition for something like these will be ok
```java
message = incomingMessages.poll(0, TimeUnit.SECONDS);

if (message == null && numMessagesInQueue() == 0) {
    pendingReceives.add(result);
} else {
    resumeReceivingFromPausedConsumersIfNeeded();
    result.complete(message);
}
```

what do you think?

Reply:
> `if (message == null && numMessagesInQueue() == 0) {..}`

I am not sure how this condition will help. If `message=null` and `numMessagesInQueue` is empty then also we want to add listener on internal individual consumer  to receive message when it's available. In your condition: if `message=null` and  `numMessagesInQueue` is not empty then it will complete result with `null` result and it will give NPE to client.

Reply:
Yes, that's correct, in that case `resumeReceivingFromPausedConsumersIfNeeded` should be able to return some Message because at that point we know for sure there is a message available in some consumer.

Reply:
Is this part of the commit needed to address the duplicates after the timeout? Or can this portion be split into a separate change? 

My concern here is that `messageFromConsumerImpl()` can be called many times and if it's checking all the partitions it might take a while.

Reply:
Not at all, I've just leave it int order to ask about these behaviour. In fact I believe there is some bug or race condition in which messages are stored at ConsumerImpl level instead of PartitionedConsumer, anyway is not a critical bug because UnackMessageTracker will redeliver those messages, but that change deservers its own issue/pr. 
Removing this code from this pr. 👍